src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs

```diff
@@ -182,29 +204,25 @@ public IEnumerable<string> GetTopLevelDependencies()
             return privateAssetsToExclude;
         }
         private static HashSet<string> GetIntersection(
-          IDictionary<string, LockFileTargetLibrary> collection1,
-          IDictionary<string, LockFileTargetLibrary> collection2)
+          IDictionary<string, HashSet<PackageIdentity>> packagesToBeFiltered,
+          IDictionary<string, LockFileTargetLibrary> packagesToBePublished)
         {
-            var exclusionList = new HashSet<string>();
-            var iterated = collection1;
-            var lookup = collection2;
-
-            if (collection1.Count > collection2.Count)
-            {
-                iterated = collection2;
-                lookup = collection1;
-            }
-            foreach (var entry in iterated)
+            var exclusionList = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            
+            foreach (var entry in packagesToBePublished)
             {
-                LockFileTargetLibrary library = lookup[entry.Key];
-
-                if (library != null)
+                var entryKey = entry.Key.ToLower();
+                if (packagesToBeFiltered.ContainsKey(entryKey))
                 {
+                    var librarySet = packagesToBeFiltered[entryKey];
                     LockFileTargetLibrary dependency = entry.Value;
-
-                    if (library.Version.Equals(dependency.Version))
+                    foreach (var library in librarySet)
                     {
-                        exclusionList.Add(entry.Key);
+                        if (dependency.Version.Equals(library.Version))
```

Comment:
@piotrpMSFT had a really good idea that we log some sort of warning/info here telling people that they are using a higher version than what is in the cache.  That way they are aware that they will not be running against the optimized version of their dependency, and as a result their app may have slower startup times.